### PR TITLE
Support for more method modifiers

### PR DESCRIPTION
**Pull request by danielmarbach**
Support for more method scenarios for #112

### CODE REVIEW

File: src/PublicApiGenerator/MethodNameBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                // too simple, improve
+                isNew = typeDef?.Methods.Any(e => e.Name.Equals(methodDefinition.Name, StringComparison.Ordinal) && e.Parameters.Count == methodDefinition.Parameters.Count);
+                if (isNew is true)
+                {
+                    break;
+                }
+                baseType = typeDef?.BaseType;
+            }
+
+            return (attributes & MemberAttributes.ScopeMask, isNew, methodDefinition.IsVirtual,
```
Comment (sungam3r) [inline src/PublicApiGenerator/MethodNameBuilder.cs]: it looks like it was copied from another file. Is it worth to extract in a method?

File: src/PublicApiGenerator/CodeNormalizer.cs [hunk 2]
```diff
@@ -83,6 +90,15 @@ static string EventModifierMatcher(Match group)
                 .Replace($"{visibility} event ", replacementBuilder.ToString());
         }
 
+        static string MethodModifierMatcher(Match group)
```
Comment (jnm2) [inline src/PublicApiGenerator/CodeNormalizer.cs]:
```suggestion
        static string MethodModifierMatcher(Match match)
```

File: src/PublicApiGenerator/MethodNameBuilder.cs (hunk 1 above)
Comment (danielmarbach) [reply src/PublicApiGenerator/MethodNameBuilder.cs]: I'm not a huge fan of generalizing too early. I'd wait and see how the property cases look like and then decide

File: src/PublicApiGenerator/MethodNameBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        public static string AugmentMethodNameWithMethodModifierMarkerTemplate(MethodDefinition methodDefinition,
+            MemberAttributes attributes)
+        {
+            var name = CSharpOperatorKeyword.Get(methodDefinition.Name);
+
+            if (methodDefinition.DeclaringType.IsInterface)
+            {
+                return name;
+            }
```
Comment (jnm2) [inline src/PublicApiGenerator/MethodNameBuilder.cs]: ℹ This is going to have to change in the future to support the many modifiers now allowed on interface members in C#.
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#default-interface-methods

File: src/PublicApiGenerator/MethodNameBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+            bool? isNew = null;
+            var baseType = methodDefinition.DeclaringType.BaseType;
+            while (baseType != null)
+            {
+                var typeDef = baseType as TypeDefinition;
+                // too simple, improve
+                isNew = typeDef?.Methods.Any(e => e.Name.Equals(methodDefinition.Name, StringComparison.Ordinal) && e.Parameters.Count == methodDefinition.Parameters.Count);
```
Comment (jnm2) [inline src/PublicApiGenerator/MethodNameBuilder.cs]: `e.Name.Equals(methodDefinition.Name, StringComparison.Ordinal)` does the same comparison as `e.Name == methodDefinition.Name`

File: src/PublicApiGenerator/MethodNameBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            if (methodDefinition.DeclaringType.IsInterface)
+            {
+                return name;
+            }
+
+            bool? isNew = null;
+            var baseType = methodDefinition.DeclaringType.BaseType;
+            while (baseType != null)
+            {
+                var typeDef = baseType as TypeDefinition;
```
Comment (jnm2) [inline src/PublicApiGenerator/MethodNameBuilder.cs]: If desired:

```cs
            while (baseType is TypeDefinition typeDef)
            {
```

File: src/PublicApiGeneratorTests/Method_modifiers.cs [hunk 6]
```diff
@@ -20,15 +48,57 @@ public class ClassWithStaticMethod
...
+        [Fact]
+        public void Should_output_static_new_modifier()
+        {
+            AssertPublicApi<ClassWithStaticNewMethod>(
+                @"namespace PublicApiGeneratorTests.Examples
+{
+    public class ClassWithStaticNewMethod : PublicApiGeneratorTests.Examples.ClassWithStaticMethod
+    {
+        public ClassWithStaticNewMethod() { }
+        public new static void DoSomething() { }
```
Comment (jnm2) [inline src/PublicApiGeneratorTests/Method_modifiers.cs]: I was thinking this was typically `static new` but I'm not finding a canonical list, so...?

Comment (danielmarbach) [reply src/PublicApiGeneratorTests/Method_modifiers.cs]: Honestly I used ReSharper suggestion but will double check

File: src/PublicApiGenerator/MethodNameBuilder.cs (hunk 3 above)
Comment (danielmarbach) [reply src/PublicApiGenerator/MethodNameBuilder.cs]: Don't get me started... 😂
Yeah but we can focus on that in another major release

File: src/PublicApiGeneratorTests/Method_modifiers.cs (hunk 6 above)
Comment (jnm2) [reply src/PublicApiGeneratorTests/Method_modifiers.cs]: Sounds good as-is 👍

Comment (danielmarbach) [reply src/PublicApiGeneratorTests/Method_modifiers.cs]: Seems microsoft mostly uses "static new"

https://github.com/dotnet/corefx/search?q=%22static+new%22&unscoped_q=%22static+new%22

I've asked on twitter. In the meantime let's pull this in as is. We can change it later once we know more

**Pull request merged**